src/Meditation.UI/Services/Dialogs/DialogLifetime.cs

```diff
@@ -8,36 +9,42 @@ public class DialogLifetime : IDisposable
     {
         private readonly Window _ownerDialog;
         private readonly Window _dialogWindow;
+        private TaskCompletionSource? _taskCompletionSource;
         private bool _isShown;
         private bool _isDisposed;
 
+        public object? DataContext => _dialogWindow?.DataContext;
+
         public DialogLifetime(Type dialogType, Window ownerDialog, IServiceProvider serviceProvider)
         {
-            if (!typeof(Window).IsAssignableFrom(dialogType))
-                throw new ArgumentException($"Supplied type {dialogType} must be a Window");
-
-            if (ActivatorUtilities.CreateInstance(serviceProvider, dialogType) is not Window dialogWindow)
-                throw new ArgumentException($"Could not construct dialog {dialogType}");
-
+            var dialogWindow = CreateDialog(dialogType, serviceProvider);
             _dialogWindow = dialogWindow;
             _ownerDialog = ownerDialog;
         }
 
-        public void Show()
+        public Task Show()
         {
             _dialogWindow.Show(_ownerDialog);
+            _taskCompletionSource = new TaskCompletionSource();
             _isShown = true;
+            return _taskCompletionSource.Task;
         }
 
         public void Close()
         {
-            if (!_isShown)
-                throw new InvalidOperationException("Dialog is not opened");
+            if (!_isShown || _taskCompletionSource == null)
```

Comment:
[Note] Typically, methods like `.Close` or `.Delete` are written in such a way that they do not throw an error if the entity is already closed/deleted. This makes them easier to use in `try-finally` blocks and reduces the number of fails in situations where nothing bad happened.